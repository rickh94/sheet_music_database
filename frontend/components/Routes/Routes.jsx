import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Switch, Route, BrowserRouter as Router } from 'react-router-dom'
import { connect } from 'react-redux'
import { account } from '../../actions'
import Loadable from 'react-loadable'

import Home from '../Home'

const Login = Loadable({
  loader: () => import('../Login'),
  loading: () => <div>Loading...</div>
})
const Logout = Loadable({
  loader: () => import('../Logout'),
  loading: () => <div>Loading...</div>
})
const NotFound = Loadable({
  loader: () => import('../NotFound'),
  loading: () => <div>Loading...</div>
})
const Profile = Loadable({
  loader: () => import('../Profile'),
  loading: () => <div>Loading...</div>
})
const Register = Loadable({
  loader: () => import('../Register'),
  loading: () => <div>Loading...</div>
})
const Tags = Loadable({
  loader: () => import('../Tags'),
  loading: () => <div>Loading...</div>
})
const Composers = Loadable({
  loader: () => import('../Composers'),
  loading: () => <div>Loading...</div>
})
const Composer = Loadable({
  loader: () => import('../Composer'),
  loading: () => <div>Loading...</div>
})


export class Routes extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isAuthenticating: true
    }
  }

  static propTypes = {
    attemptLogin: PropTypes.func.isRequired
  }

  async componentDidMount() {
    const token = localStorage.getItem('token')
    if (token) {
      await this.props.attemptLogin(token)
    }
    this.setState({ isAuthenticating: false })
  }

  render() {
    return (
      !this.state.isAuthenticating && (
        <Router>
          <Switch>
            <Route exact path="/" component={Home} />
            <Route exact path="/login" component={Login} />
            <Route exact path="/logout" component={Logout} />
            <Route exact path="/register" component={Register} />
            <Route exact path="/profile" component={Profile} />
            <Route exact path="/tags" component={Tags} />
            <Route exact path="/composers" component={Composers} />
            <Route exact path="/composer/:id" component={Composer} />
            <Route default component={NotFound} />
          </Switch>
        </Router>
      )
    )
  }
}

const mapDispatchToProps = dispatch => {
  return {
    attemptLogin: token => {
      return dispatch(account.checkToken(token))
    }
  }
}

export default connect(
  undefined,
  mapDispatchToProps
)(Routes)
