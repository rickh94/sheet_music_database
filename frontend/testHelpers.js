import React from 'react'

export const testToken = 'testtoken'

export function testLoggedIn(wrapper) {
  const history = { push: jest.fn() }
  const token = null
  const alert = { show: jest.fn() }
  wrapper.setProps({ token, history, alert })
  wrapper.instance().componentDidMount()
  expect(history.push).toHaveBeenCalledWith('/login')
  expect(alert.show).toHaveBeenCalledWith(
    <span className="alert-text">Please log in</span>,
    { type: 'error' }
  )
}

export function testGetData(wrapper, functionPropName, functionPropArgs = []) {
  const getFunction = jest.fn()
  wrapper.setProps({ [functionPropName]: getFunction, token: testToken })
  wrapper.instance().componentDidMount()
  expect(getFunction).toHaveBeenCalledWith(testToken, ...functionPropArgs)
}

export function testGetDataFailed(wrapper, functionPropName, nameInMessage) {
  const getFunction = () => false
  const alert = { show: jest.fn() }
  wrapper.setProps({ [functionPropName]: getFunction, token: testToken, alert })
  wrapper.instance().componentDidMount()
  expect(alert.show).toHaveBeenCalledWith(
    <span className="alert-text">{`Could not get ${nameInMessage}`}</span>,
    { type: 'error' }
  )
}

export function clickThing(tag, wrapper, content) {
  wrapper
    .findWhere(el => el.name() == tag && el.contains(content))
    .simulate('click', { preventDefault: jest.fn() })
}

export const clickButton = (wrapper, content) => clickThing('Button', wrapper, content)
export const clickLink = (wrapper, content) => clickThing('a', wrapper, content)

