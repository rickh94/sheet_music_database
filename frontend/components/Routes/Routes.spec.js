import React from 'react'
import { shallow } from 'enzyme'
import Routes from './Routes'

import { testToken } from '../../testHelpers'

import '../../setupTests'

describe('Routes', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<Routes attemptLogin={jest.fn()} />)
  })

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a switch when it finishes trying to authenticate', () => {
    wrapper.setState({ isAuthenticating: false })
    expect(wrapper.exists('Switch')).toBeTruthy()
  })

  it('attempts to login from localStorage token and unset isAuthenticating', async () => {
    localStorage.setItem('token', testToken)
    const attemptLogin = jest.fn()
    wrapper.setProps({ attemptLogin })
    await wrapper.instance().componentDidMount()
    expect(attemptLogin).toHaveBeenCalledWith(testToken)
    expect(wrapper.state().isAuthenticating).toBeFalsy()
  })

  it('does not attempt login without localstorage token', async () => {
    localStorage.removeItem('token')
    const attemptLogin = jest.fn()
    wrapper.setProps({ attemptLogin })
    await wrapper.instance().componentDidMount()
    expect(attemptLogin).not.toHaveBeenCalledWith()
    expect(wrapper.state().isAuthenticating).toBeFalsy()
  })
})
