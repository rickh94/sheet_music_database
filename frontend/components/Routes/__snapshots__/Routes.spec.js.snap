// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Routes
    attemptLogin={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Switch>
        <Route
          component={[Function]}
          exact={true}
          path="/"
        />
        <Route
          component={[Function]}
          exact={true}
          path="/login"
        />
        <Route
          component={[Function]}
          exact={true}
          path="/logout"
        />
        <Route
          component={[Function]}
          exact={true}
          path="/register"
        />
        <Route
          component={[Function]}
          exact={true}
          path="/profile"
        />
        <Route
          component={[Function]}
          exact={true}
          path="/tags"
        />
        <Route
          component={[Function]}
          exact={true}
          path="/composers"
        />
        <Route
          component={[Function]}
          exact={true}
          path="/composer/:id"
        />
        <Route
          component={[Function]}
          default={true}
        />
      </Switch>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Route
            component={[Function]}
            exact={true}
            path="/"
          />,
          <Route
            component={[Function]}
            exact={true}
            path="/login"
          />,
          <Route
            component={[Function]}
            exact={true}
            path="/logout"
          />,
          <Route
            component={[Function]}
            exact={true}
            path="/register"
          />,
          <Route
            component={[Function]}
            exact={true}
            path="/profile"
          />,
          <Route
            component={[Function]}
            exact={true}
            path="/tags"
          />,
          <Route
            component={[Function]}
            exact={true}
            path="/composers"
          />,
          <Route
            component={[Function]}
            exact={true}
            path="/composer/:id"
          />,
          <Route
            component={[Function]}
            default={true}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/login",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/logout",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/register",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/profile",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/tags",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/composers",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "exact": true,
            "path": "/composer/:id",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "default": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Switch>
          <Route
            component={[Function]}
            exact={true}
            path="/"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/login"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/logout"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/register"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/profile"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/tags"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/composers"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/composer/:id"
          />
          <Route
            component={[Function]}
            default={true}
          />
        </Switch>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Route
              component={[Function]}
              exact={true}
              path="/"
            />,
            <Route
              component={[Function]}
              exact={true}
              path="/login"
            />,
            <Route
              component={[Function]}
              exact={true}
              path="/logout"
            />,
            <Route
              component={[Function]}
              exact={true}
              path="/register"
            />,
            <Route
              component={[Function]}
              exact={true}
              path="/profile"
            />,
            <Route
              component={[Function]}
              exact={true}
              path="/tags"
            />,
            <Route
              component={[Function]}
              exact={true}
              path="/composers"
            />,
            <Route
              component={[Function]}
              exact={true}
              path="/composer/:id"
            />,
            <Route
              component={[Function]}
              default={true}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/login",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/logout",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/register",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/profile",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/tags",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/composers",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/composer/:id",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "default": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
